book/src/message_validation.md

```diff
@@ -49,12 +49,13 @@
 
 | ValidationStep | Description                                                                            | Implemented | Tested | Test File |
 | -------------- | -------------------------------------------------------------------------------------- | ----------- | ------ | --------- |
-| `ValSem200`    | Commit must not cover inline self Remove proposal                                      | ✅          | ❌     | TBD       |
-| `ValSem201`    | Path must be present, if Commit contains Removes or Updates                            | ❌          | ❌     | TBD       |
-| `ValSem202`    | Path must be the right length                                                          | ❌          | ❌     | TBD       |
-| `ValSem203`    | Path secrets must decrypt correctly                                                    | ❌          | ❌     | TBD       |
-| `ValSem204`    | Public keys from Path must be verified and match the private keys from the direct path | ✅          | ❌     | TBD       |
-| `ValSem205`    | Confirmation tag must be successfully verified                                         | ✅          | ❌     | TBD       |
+| `ValSem200`    | Commit must not cover inline self Remove proposal                                      | ✅          | ✅     | `openmls/src/group/tests/test_commit_validation.rs`       |
+| `ValSem201`    | Path must be present, if Commit is empty or contains proposals other than Add, PreSharedKey or ReInit                                       | ✅          | ✅     | `openmls/src/group/tests/test_commit_validation.rs`       |
```

Comment:
Ah, one paragraph above that (and I guess that one is authoritative), it says:
> The path field of a Commit message MUST be populated if the Commit covers at least one Update or Remove proposal. The path field MUST also be populated if the Commit covers no proposals at all (i.e., if the proposals vector is empty). The path field MAY be omitted if the Commit covers only Add proposals. In pseudocode, the logic for validating a Commit is as follows:
> ```
> hasUpdates = false
> hasRemoves = false
>
> for i, id in commit.proposals:
>    proposal = proposalCache[id]
>    assert(proposal != null)
>
>    hasUpdates = hasUpdates || proposal.msg_type == update
>    hasRemoves = hasRemoves || proposal.msg_type == remove
>
> if len(commit.proposals) == 0 || hasUpdates || hasRemoves:
>  assert(commit.path != null)
> ```

Reply:
Right, but right above that definition there's the following set of rules and the pseudo-code:

> The path field of a Commit message MUST be populated if the Commit covers at least one Update or Remove proposal. The path field MUST also be populated if the Commit covers no proposals at all (i.e., if the proposals vector is empty). The path field MAY be omitted if the Commit covers only Add proposals. In pseudocode, the logic for validating a Commit is as follows:

```
hasUpdates = false
hasRemoves = false

for i, id in commit.proposals:
    proposal = proposalCache[id]
    assert(proposal != null)

    hasUpdates = hasUpdates || proposal.msg_type == update
    hasRemoves = hasRemoves || proposal.msg_type == remove

if len(commit.proposals) == 0 || hasUpdates || hasRemoves:
  assert(commit.path != null)
  ```

Reply:
Ah, I think we got there at the same time. If you want, I can change the definition back. But I still think it's saner to define the need for a path by when it's not needed rather than the other way around and I think the spec should be changed accordingly. Do you have a strong opinion on this?

Reply:
I'm not sure I follow your thoughts on what's wrong with the spec. The spec clearly states in which cases a path is mandatory. In all other cases, the path is optional.
Assuming that's fine, we should test for all mandatory cases (but we don't need to test for other cases.

Reply:
Yes, I agree it contradicts the spec, so I'll change it back. I think it would be safer to define it the other way around, though. I opened an issue with the spec: https://github.com/mlswg/mls-protocol/issues/596